### ISSUE TITLE
Synapse fails to restart after zookeeper node disappears

### ISSUE THREAD

**Issue by Jaykah** [airbnb/synapse#94]
I am using the Zookeeper watcher to get a list of available nodes - which it successfully does.

However, after a node fails, it does not get removed from haproxy.

Haproxy monitoring page shows those nodes as down, Zookeeper no longer has them as they are ephemeral, but the reconfiguration does not take place.

If I restart Synapse, however, everything goes back to normal.

```
I, [2014-10-09T22:17:20.844370 #12]  INFO -- Synapse::ZookeeperWatcher: synapse: discovering backends for service proddb
I, [2014-10-09T22:17:20.861672 #12]  INFO -- Synapse::ZookeeperWatcher: synapse: discovered 6 backends for service proddb
[WARNING] 281/221756 (157) : Server proddb/100.0.10.10:3306_100.0.10.10 is DOWN, reason: Layer4 connection problem, info: "Connection refused", check duration: 1ms. 6 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
[WARNING] 281/221756 (157) : Server proddb/100.0.10.10:3306_100.0.10.10 is DOWN, reason: Layer4 connection problem, info: "Connection refused", check duration: 2ms. 5 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
I, [2014-10-09T22:18:15.682057 #12]  INFO -- Synapse::ZookeeperWatcher: synapse: discovering backends for service proddb
I, [2014-10-09T22:18:15.690843 #12]  INFO -- Synapse::ZookeeperWatcher: synapse: discovered 5 backends for service proddb
```

Note: Synapse is deployed in Docker and managed by supervisord

**Comment by hankya** [comment]
is this fixed? anyone?

**Comment by jolynch** [comment]
@hankya We haven't seen the behavior referenced in this PR but we don't use the supervisord. Have you been seeing the same issue and if so what's your setup?

I did see something that I thought was like this bug when ZK was overloaded, but in that case the issue was just a large lag in event delivery. Our connection pooling fix alleviated those issues (plus a bug fix to the pooling code). While investigating the pooling bug though I setup a pretty thorough test setup where I was removing and adding nodes while messing with the network via netem and I still wasnt able to repro with Ruby 1.9 and running Synapse directly. I could try to modify that setup to get a repro of this issue if it's still impacting people.

**Comment by Jaykah** [comment]
@jolynch This issue is related to the fact that (AFAIK) Synapse does reloads by forking the process and then executing the reload. The problem with Supervisord is that it keeps attached to STDOUT, running the process in the foreground, thus messing with the ability to fork&reload. That's a pity, because Supervisord is a really great tool for process monitoring.

**Comment by igor47** [comment]
i don't think that makes sense. we run synapse via runit, which also keeps the process attached in the foreground, and we have no problems. also, i'm not sure what you're referring to as far as reloads via forking. our usual model is to exit the process and allow the process supervisor to restart it.

**Comment by jolynch** [comment]
I'm also confused by the forking issue. We use upstart and just have it kill the old process before starting the new one. For nerve we run two copies during a flip to achieve hitless nerve reloads. @Jaykah why do you need to fork Synapse to reload?

I think that @hankya is having a separate issue, which I'm following up on in #137.

**Comment by Jaykah** [comment]
@jolynch The first six comments of this ticket explain my reasoning. Basically what happens is if we have only one instance of Synapse running it does not write modified Haproxy configs for node removal, if we run two it does remove nodes, but does not add new ones. And if we run is as numprocs=4 (basically forking it into 4 processes) it *seems to* work fine, but I've still had some cases with reloads not happening (maybe once every 100 reloads or so as opposed to 100% of the time without forking). Unfortunately, this is something that I wasn't yet able to figure out...

**Comment by jolynch** [comment]
@Jaykah I think the relevant context here is that you are using ruby 2.1.2, same as @hankya over in #137. I will focus my efforts on that ruby version. It's also been reported that 2.2 doesn't work (although ruby-zk doesn't work with 2.2 so ...).

For context, I am unable to reproduce this at all in our production environment or in my own testing using ruby 1.9, so if you can downgrade, that might be a good short term solution.

**Comment by jolynch** [comment]
@igor47 I thought you had mentioned airbnb moving to ruby 2.X soon, if you guys get around to debugging this before me that is also cool heh :-)

**Comment by igor47** [comment]
fyi @rickypai

**Comment by rickypai** [comment]
I sent a PR to the zk gem to have its tests run and pass against Ruby 2.2. Unfortunately, we are yet to be running our infrastructure on Ruby 2.x, so we have not seen this error.

**Comment by c4urself** [comment]
FWIW, I've run the following test on Ubuntu 14.04 which works on ruby1.9.1 and *does not* work on ruby2.0 running a supervisord based nerve and synapse.

1. [on service B] `supervisorctl stop nerve`
  * Correct. HAproxy keeps servers in backend.
2. [on service A] `supervisorctl stop synapse`
  * Correct. HAproxy keeps servers in backend.
3. [on service A] `supervisorctl start synapse`
  * Correct. Synapse falls back to default servers
4. [on service B] `supervisorctl start nerve`
   * Correct. Nerve starts fine and registers with Zookeeper on service B
   * **Incorrect. Synapse does not get callback for change and does not include newly registered servers in backend, keeps running on default servers on service A**

The difference is that on step 4, ruby1.9.1 synapse re-registers the backends.

**Comment by jolynch** [comment]
@c4urself did you run this test with latest master?  We had a bug with a Thread api change between 1.9 and 2.0, but I thought that I fixed it in PR #138.

**Comment by c4urself** [comment]
@jolynch No, I ran the test with `0.12.1`. I retried it with the master branch and it looks like it's working correctly now. Thanks!

**Comment by jolynch** [comment]
@igor47 can you cut a bug fix release? I haven't been able to roll it out here but if you've done that then we probably want to cut a release.

**Comment by igor47** [comment]
i'll do that now; i've only given it the barest testing on our infra, haven't rolled it out site-wide yet

**Comment by jolynch** [comment]
@Jaykah how does master work for you? Do you still see the lack of restart on update?

**Comment by Jaykah** [comment]
@jolynch Have just tested it, and it still does not do reloads properly. Built from master.

```
I, [2015-10-12T20:47:38.823828 #18]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovering backends for service application
I, [2015-10-12T20:47:38.855753 #18]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovered 2 backends for service application
I, [2015-10-12T20:47:38.857344 #18]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovering backends for service application
```

**Comment by jolynch** [comment]
@Jaykah, OK I'll look into it. Can you post your docker file, it might make it easier for me to reproduce.

Just to make sure we're talking about the same thing, Synapse is not supposed to reload itself if Synapse configuration changes (adding or removing services) however it is supposed to reload HAProxy if the servers backing those services change. I guess the part I'm confused about in this ticket is why Synapse needs to fork at all to reload HAProxy.

Actually, how does this work at all in a Docker container? Does HAProxy run inside the container?

**Comment by Jaykah** [comment]
@jolynch Sure, will post the Dockerfile shortly.

That is correct, I am referring to the HAProxy reload when the list of nodes changes in Zookeeper. So the steps would be:

1) Launch a new backend container that has Nerve
2) It will register its IP in Zookeeper
3) Frontend container's Synapse would see that the list of nodes changed in Zk
4) It should then reload HAProxy (but doesn't).

Both HAProxy and Synapse run in Docker with Supervisord.

**Comment by igor47** [comment]
it needs to fork to reload haproxy because the haproxy reload is accomplished with a command. the unix way to execute another command is to `fork` and then `execve`. i wonder if this is docker-related at this point

**Comment by jolynch** [comment]
I think that having the docker file (or something close to it) will be a big help in reproducing the bug. I have a suspicion supervisord and docker are not acting like we might expect.

**Comment by jolynch** [comment]
Yea there are going to be two forks (one by Synapse and one by the HAProxy that starts up), so if supervisord is like "what is this thing that just forked" that might be an issue.

**Comment by Jaykah** [comment]
@jolynch I've created a Dockerfile bundle for you that can be used to reproduce the issue. All you have to do is add a "synapse.conf.yaml" and your public key (to ssh into container via ip:222), named "myserver.pub".

You will then be able to connect to Supervisord console via "supervisorctl" with user "supervisord" and password "testpass". Port 3212 is also exposed for HAProxy stats.

https://www.dropbox.com/sh/yjs6eds1t0xdpg7/AAA7BUQ-MkP8nfOyjPvAZ549a?dl=0

**Comment by Jaykah** [comment]
@jolynch Were you able to reproduce it?

**Comment by jolynch** [comment]
@Jaykah sorry I'll try to take a look this weekend, sorry about that.

**Comment by Jaykah** [comment]
@jolynch Let me know if I can help in solving that issue in any way, would be glad to provide a staging environment or any other information that might be helpful.

**Comment by jolynch** [comment]
@Jaykah I'm still working on getting a repro working, but one of the first things I noticed is that you have the supervisord process managing haproxy which is probably not going to work with Synapse causing it to fork. This sounds similar to the problems that folks have with systemsd which assumes that the process it's managing does not fork. See [the patch](http://marc.info/?l=haproxy&m=136033560505600&w=2) for more background.

So I guess two questions:
1. What is your haproxy reload command in your synapse.conf.json?
2. Have you tried managing the haproxy-systemd-wrapper (instead of haproxy) with supervisord and then having synapse trigger a SIGHUP2, see [this post](http://marc.info/?l=haproxy&m=142177973225397&w=2) for some insight.

**Comment by Jaykah** [comment]
@jolynch 

```
[program:haproxy] 
command=/usr/sbin/haproxy -db -f /etc/haproxy/haproxy.cfg
```

Haproxy is launched in foreground as 1 process, but then exits with a status 0 (expected) when Synapse loads it. That way we are monitoring the initial launch of Haproxy without restraining forking (for the purpose you've mentioned). We are then stopping it's management by Supervisord with:
```
autorestart=unexpected
```

1. The reload command is the following:

```reload_command: "haproxy -db -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -sf $(pidof haproxy)"```

2. Will try that shortly.

**Comment by Jaykah** [comment]
@jolynch Unfortunately nothing changes by using the wrapper. Synapse is not restarting Haproxy:

```
I, [2015-11-11T19:27:53.419479 #345]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovering backends for service application
I, [2015-11-11T19:27:53.423973 #345]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovered 1 backends for service application
I, [2015-11-11T19:27:53.969748 #345]  INFO -- Synapse::Synapse: synapse: configuring haproxy
I, [2015-11-11T19:27:53.971475 #345]  INFO -- Synapse::Haproxy: synapse: reconfigured haproxy
```

If I restart Synapse manually I get:

```
I, [2015-11-11T19:29:36.421316 #366]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovering backends for service proddb
I, [2015-11-11T19:29:36.429431 #366]  INFO -- Synapse::ServiceWatcher::ZookeeperWatcher: synapse: discovered 4 backends for service proddb
I, [2015-11-11T19:29:36.429514 #366]  INFO -- Synapse::Synapse: synapse: configuring haproxy
I, [2015-11-11T19:29:36.431798 #366]  INFO -- Synapse::Haproxy: synapse: reconfigured haproxy
I, [2015-11-11T19:29:36.450096 #366]  INFO -- Synapse::Haproxy: synapse: restarted haproxy
```

Moreover, with supervisord.conf
```
command=/usr/sbin/haproxy-systemd-wrapper -Ds -f /etc/haproxy/haproxy.cfg
```
and synapse.conf.yaml
```
reload_command: "/bin/kill -SIGUSR2 $(pidof haproxy-systemd-wrapper)"
```

I can still see Haproxy's forks even after the process is terminated via Supervisor.



Were you able to reproduce the issue with the Dockerfile I provided?

**Comment by Jaykah** [comment]
Did some more tests with/without -Ds, with/without -db, with numprocs=1/2/3/4, without fork control, restarts/reloads via -USR2, -HUP - nothing worked. At this time I am almost certain that the issue has to do with Supervisor reaping Synapse processes because Synapse does not have a "foreground" (or non-daemon) mode, making it not systemd compliant.

```
2015-11-13 02:37:41,657 CRIT reaped unknown pid 478)
2015-11-13 02:38:46,552 CRIT reaped unknown pid 473)
2015-11-13 02:38:46,562 CRIT reaped unknown pid 477)
2015-11-13 02:38:46,564 CRIT reaped unknown pid 479)
```

**Comment by Jaykah** [comment]
@jolynch Were you able to reproduce it? We're debating if we should move our infra off Synapse because of this issue, but if you have any pointers as to how we can make it run in foreground, that would be super helpful.

Thanks.

**Comment by jolynch** [comment]
@Jaykah Yea I've been dropping the ball on this. I really think that the issue lies with how a Docker container really isn't a proper init system. We really like running synapse/haproxy outside the containers because of all kinds of reasons (namely our service data plane stays available during a docker roll and we don't like paying conntrack), is there any possibility you can run on the bare metal? To be honest I thought that Synapse does stay attached in the foreground (as @igor47 has implied they run it via runit).

Why do you think those reaps are Synapse PIDs? Couldn't they just as easily be HAProxy instances? 

I'll keep working on repro but ... well I've been saying that for a while.

**Comment by jolynch** [comment]
@Jaykah Alright I got a few hours to debug this today and this is not a bug in Synapse, it's an issue with how supervisord and haproxy interact. There are a couple of problems with running Synapse like this but they all revolve around the problem that Supervisord assumes that it is the _only_ thing creating processes ever and this is simply not true with Synapse (which typically forks haproxy) or HAProxy (which typically forks itself).

Because we're running in supervisord the only way that Synapse can reload the service is via supervisorctl, e.g. ``supervisorctl restart haproxy:haproxy`` because it can't just fork off a process as we usually would because supervisord can't handle that (this leads to the reaping of unknown pid problem you were mentioning).

However, if we do that and use the haproxy-systemd-wrapper, then we run into the problem that supervisord only supports stop then start restarts (I think), which means that it waits for the previous haproxy to completely finish before it will start the new one. This is bad because the way that HAProxy gracefully drains traffic is to have the old process drop its listeners but crucially _not exit_ until all TCP connections are finished cleanly. This leads to the unknown pid problem as well, because there are a bunch of haproxy instances hanging around that will never go away due to the way the haproxy systemd wrapper works.

I _think_ this would all be solved if we could just supply the pid in the ``command`` for HAProxy but I don't know how to do this in supervisord.

I'm going to keep poking at this to see if I can properly work around supervisord and haproxy's terribleness, but at this point I'm pretty convinced that this is not a bug with Synapse but rather an issue with supervisord being a really bad replacement for a proper init system.

**Comment by jolynch** [comment]
For what it's worth I was able to get a container that would restart haproxy, but it worked by basically having a bash script that would exec:

```bash
#!/bin/bash                                                                                                                    
touch /var/haproxy/haproxy.pid                                                  
exec /usr/sbin/haproxy -db -f /etc/haproxy/haproxy.cfg -p /var/haproxy/haproxy.pid -sf $(cat /var/haproxy/haproxy.pid)
```

And then a supervisord configuration that would just use this as the command and ``USR2`` as the stopsignal. This is really bad though because it means there will be a dead period where no HAProxy is listening for connections between when supervisord stops the old and HAProxy and starts the new one.

I'm pulling in HAProxy 1.6 to see if they've made any improvements to the haproxy-systemd-wrapper, but given my reading of the source code the haproxy-systemd-wrapper I'm not hopeful.

@Jaykah I really think that this is going to work better running on bare metal... but it would be nice to support a docker deployment so I'll keep poking.

**Comment by jolynch** [comment]
Alright, I've got a working Docker setup with latest synapse, haproxy 1.6 and ruby 2.2 over at https://gist.github.com/jolynch/54db6eb64a8951450d6e. The TLDR is: use the systemd wrapper, have synapse send signals to the systemd wrapper (not to haproxy itself) and have supervisord always autostart the wrapper and try to start it before synapse (there is a bit of a chicken and an egg problem because synapse has to generate the config before supervisord can start it properly).

But yea, I really wouldn't encourage someone doing this in production because:

1. Your service data plane (haproxy) is tied to this docker image's uptime. This means that you can't restart synapse with new configuration or reload docker without dropping traffic.
2. HAProxy really plays a _lot_ better with real init systems, it is designed for bare metal (and it's packaged on basically every platform so this isn't a huge deal imo...)
3. I don't think that Docker actually buys you that much over building .deb/.rpm/w.e. packages your distro provides. I think the only dependency that's tricky is ruby, and that's really not that tricky.
